parsec/core/gui/trio_thread.py

```diff
@@ -247,71 +174,34 @@ async def _throttled_execute(task_status=trio.TASK_STATUS_IGNORED):
             # Actually start the job execution
             self._throttling_last_executed[throttling_id] = trio.current_time()
             try:
-                await self._send_job_channel.send(job)
+                await self.nursery.start(job)
             except trio.BrokenResourceError:
                 # Job scheduler has been closed, nothing more can be done
                 pass
 
-        async def _send_throttle_job():
-            # While inside this async function we are blocking the Qt thread
-            # hence we just submit `_throttled_execute`
-            await self._send_job_channel.send(_throttled_execute)
-
-        try:
-            self._run_job(_send_throttle_job)
-        except JobSchedulerNotAvailable:
-            pass
+        # Create the job but don't execute it: we have to handle throttle first !
+        job = QtToTrioJob(fn, args, kwargs, qt_on_success, qt_on_error)
+        if self.nursery._closed:
+            job.set_cancelled(JobSchedulerNotAvailable())
+        else:
+            self.nursery.start_soon(_throttled_execute, job)
+        return job
 
     def submit_job(self, qt_on_success, qt_on_error, fn, *args, **kwargs):
-        # Fool-proof sanity check, signals must be wrapped in `ThreadSafeQtSignal`
-        assert not [x for x in args if isinstance(x, pyqtBoundSignal)]
-        assert not [v for v in kwargs.values() if isinstance(v, pyqtBoundSignal)]
-        job = QtToTrioJob(self._trio_token, fn, args, kwargs, qt_on_success, qt_on_error)
-
-        async def _submit_job():
-            # While inside this async function we are blocking the Qt thread
-            # hence we just wait for the job to start (to avoid concurrent
-            # crash if the job is cancelled)
-            await self._send_job_channel.send(job)
-            await job._started.wait()
-
-        try:
-            self._run_job(_submit_job)
-        except JobSchedulerNotAvailable as exc:
-            job.set_cancelled(exc)
-
+        job = QtToTrioJob(fn, args, kwargs, qt_on_success, qt_on_error)
+        if self.nursery._closed:
```

Comment:
shouldn't `self.nursery.start_soon` already raises a dedicated exception if the nursery is closed ?

Reply:
Yes but it's a `RuntimeError`, and I wasn't comfortable catching all runtime errors.